### PR TITLE
refactor(snap): Combine install and configure hook executables

### PR DESCRIPTION
**Pull request by MonicaisHer**
**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-virtual-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [ ] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?)
- [ ] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

### CODE REVIEW

**Comment by farshidtz** [comment]
The application name can be made more generic. In edgex-go we have runtime helpers written in go misusing a hook: https://github.com/edgexfoundry/edgex-go/blob/main/snap/local/runtime-helpers/bin/security-proxy-post-setup.sh

Proposals:
- runtime-helper
- hook-helper
- helper -> rename runtime-helpers to helper and keep startup-env-var.sh in the same directory with go code?
```
./helpers install
./helpers configure
./helpers configure-options --service=secrets-config
```


- helper-go (I like this one)
```
./helper-go install
./helper-go configure
./helper-go configure-options --service=secrets-config
```

File: snap/local/hooks/Makefile [hunk 1]
```diff
@@ -1,14 +1,5 @@
...
-
-build: $(HOOKS)
-
-cmd/configure/configure:
-	$(GO) build $(GOFLAGS) -o $@ ./cmd/configure
-
-cmd/install/install:
-	$(GO) build $(GOFLAGS) -o $@ ./cmd/install
+build:
+	go build -o helper-go
```
Comment (farshidtz) [inline snap/local/hooks/Makefile]: The ldflags should be included.

File: snap/snapcraft.yaml [hunk 2]
```diff
@@ -26,6 +26,7 @@ apps:
     command: bin/device-virtual $CONFIG_PRO_ARG $CONF_ARG $REGISTRY_ARG
     command-chain:
       - bin/startup-env-var.sh
+      - bin/helper-go.sh
```
Comment (farshidtz) [inline snap/snapcraft.yaml]: Why is this added in the command chain?

File: snap/local/runtime-helpers/bin/helper-go.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash -e
+
+$SNAP/snap/hooks/helper-go install
+$SNAP/snap/hooks/helper-go configure
```
Comment (farshidtz) [inline snap/local/runtime-helpers/bin/helper-go.sh]: This script should be split onto two. One for install and one for configure. The scripts should be placed under snap/hooks.

Snapd calls snap/hooks/configure and snap/hooks/configure executables. We no longer have them as binaries so we instead have to place the scripts. Refer to https://snapcraft.io/docs/supported-snap-hooks

File: snap/local/hooks/helper-go.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	subCommand := os.Args[1]
+	switch subCommand {
+	case "install":
+		install()
+	case "configure":
+		configure()
+	default:
+		panic("Unknown hook sub-command: " + subCommand)
+	}
+}
```
Comment (farshidtz) [inline snap/local/hooks/helper-go.go]: This script should be at snap/local/helper-go/main.go together with install.go and configure.go.

File: snap/snapcraft.yaml [hunk 5]
```diff
@@ -45,8 +46,7 @@ parts:
     override-build: |
       cd $SNAPCRAFT_PART_SRC
       make build
-      install -DT ./cmd/configure/configure $SNAPCRAFT_PART_INSTALL/snap/hooks/configure
-      install -DT ./cmd/install/install $SNAPCRAFT_PART_INSTALL/snap/hooks/install
+      install -DT ./helper-go $SNAPCRAFT_PART_INSTALL/snap/hooks/helper-go
```
Comment (farshidtz) [inline snap/snapcraft.yaml]: Install the binary at `$SNAPCRAFT_PART_INSTALL/bin/helper-go`

**Comment by MonicaisHer** [comment]
There is a snap options related error: 
```
sudo snap install edgex-device-virtual_2.2.0-dev.7_amd64.snap --dangerous
error: cannot perform the following tasks:
- Run configure hook of "edgex-device-virtual" snap if present (run hook "configure": edgex-device-virtual.configure: could not process options: failed to write /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp  - open /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp: no such file or directory)
```

error comes from: https://github.com/canonical/edgex-snap-hooks/blob/main/options/env.go#L91

**Comment by farshidtz** [comment]
> There is a snap options related error:
> 
> ```
> sudo snap install edgex-device-virtual_2.2.0-dev.7_amd64.snap --dangerous
> error: cannot perform the following tasks:
> - Run configure hook of "edgex-device-virtual" snap if present (run hook "configure": edgex-device-virtual.configure: could not process options: failed to write /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp  - open /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp: no such file or directory)
> ```
> 
> error comes from: https://github.com/canonical/edgex-snap-hooks/blob/main/options/env.go#L91

Maybe set DEBUG=true and look at the syslogs?

**Comment by MonicaisHer** [comment]
> > There is a snap options related error:
> > ```
> > sudo snap install edgex-device-virtual_2.2.0-dev.7_amd64.snap --dangerous
> > error: cannot perform the following tasks:
> > - Run configure hook of "edgex-device-virtual" snap if present (run hook "configure": edgex-device-virtual.configure: could not process options: failed to write /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp  - open /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp: no such file or directory)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > error comes from: https://github.com/canonical/edgex-snap-hooks/blob/main/options/env.go#L91
> 
> Maybe set DEBUG=true and look at the syslogs?

Do you mean setting edgex-device-virtual's `loglevel=DEBUG`? It doesn't show any debug logs because the snap does not install successfully. However, I get some syslogs:
```
Jun 22 18:25:16 ubuntu systemd[1]: Started snap.edgex-device-virtual.hook.configure.c5a28e71-f9e5-4eb0-bc37-dcf1c9003ce7.scope.
Jun 22 18:25:17 ubuntu edgex-device-virtual.configure[99844]: Enabling config options
Jun 22 18:25:17 ubuntu edgex-device-virtual.configure[99844]: Processing options
Jun 22 18:25:17 ubuntu edgex-device-virtual.configure[99844]: Processing app options: true
Jun 22 18:25:17 ubuntu edgex-device-virtual.configure[99844]: Writing to env file /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env:
Jun 22 18:25:17 ubuntu edgex-device-virtual.configure[99844]: could not process options: failed to write /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp  - open /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp: no such file or directory
Jun 22 18:25:17 ubuntu systemd[1]: snap.edgex-device-virtual.hook.configure.c5a28e71-f9e5-4eb0-bc37-dcf1c9003ce7.scope: Succeeded.
Jun 22 18:25:17 ubuntu snapd[993]: taskrunner.go:271: [change 9727 "Run configure hook of \"edgex-device-virtual\" snap if present" task] failed: run hook "configure": edgex-device-virtual.configure: could not process options: failed to write /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp  - open /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp: no such file or directory
Jun 22 18:25:17 ubuntu snapd[993]: services.go:919: RemoveSnapServices - disabling snap.edgex-device-virtual.device-virtual.service
```

**Comment by farshidtz** [comment]
> > > There is a snap options related error:
> > > ```
> > > sudo snap install edgex-device-virtual_2.2.0-dev.7_amd64.snap --dangerous
> > > error: cannot perform the following tasks:
> > > - Run configure hook of "edgex-device-virtual" snap if present (run hook "configure": edgex-device-virtual.configure: could not process options: failed to write /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp  - open /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp: no such file or directory)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > error comes from: https://github.com/canonical/edgex-snap-hooks/blob/main/options/env.go#L91
> > 
> > 
> > Maybe set DEBUG=true and look at the syslogs?
> 
> Do you mean setting edgex-device-virtual's `loglevel=DEBUG`? It doesn't show any debug logs because the snap does not install successfully. However, I get some syslogs:
> 
> ```
> Jun 22 18:25:16 ubuntu systemd[1]: Started snap.edgex-device-virtual.hook.configure.c5a28e71-f9e5-4eb0-bc37-dcf1c9003ce7.scope.
> Jun 22 18:25:17 ubuntu edgex-device-virtual.configure[99844]: Enabling config options
> Jun 22 18:25:17 ubuntu edgex-device-virtual.configure[99844]: Processing options
> Jun 22 18:25:17 ubuntu edgex-device-virtual.configure[99844]: Processing app options: true
> Jun 22 18:25:17 ubuntu edgex-device-virtual.configure[99844]: Writing to env file /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env:
> Jun 22 18:25:17 ubuntu edgex-device-virtual.configure[99844]: could not process options: failed to write /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp  - open /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp: no such file or directory
> Jun 22 18:25:17 ubuntu systemd[1]: snap.edgex-device-virtual.hook.configure.c5a28e71-f9e5-4eb0-bc37-dcf1c9003ce7.scope: Succeeded.
> Jun 22 18:25:17 ubuntu snapd[993]: taskrunner.go:271: [change 9727 "Run configure hook of \"edgex-device-virtual\" snap if present" task] failed: run hook "configure": edgex-device-virtual.configure: could not process options: failed to write /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp  - open /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp: no such file or directory
> Jun 22 18:25:17 ubuntu snapd[993]: services.go:919: RemoveSnapServices - disabling snap.edgex-device-virtual.device-virtual.service
> ```

No I meant the snap hook's `debug`: https://github.com/canonical/edgex-snap-hooks/blob/main/log/init.go#L10

**Comment by farshidtz** [comment]
> There is a snap options related error:
> 
> ```
> sudo snap install edgex-device-virtual_2.2.0-dev.7_amd64.snap --dangerous
> error: cannot perform the following tasks:
> - Run configure hook of "edgex-device-virtual" snap if present (run hook "configure": edgex-device-virtual.configure: could not process options: failed to write /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp  - open /var/snap/edgex-device-virtual/x3/config/device-virtual/res/device-virtual.env.tmp: no such file or directory)
> ```
> 
> error comes from: https://github.com/canonical/edgex-snap-hooks/blob/main/options/env.go#L91

This error doesn't seem to happen in CI tests. Perhaps you have some snap option set locally triggering an unknown bug.

The install hook which is called before the configure hook should have already created the needed directory. Maybe the install hook isn't called or isn't doing it's job. Or maybe an MkdirAll is missing and the directory structure isn't ready when the snap is being installed and while it has some snap options set.

**Comment by MonicaisHer** [comment]
@farshidtz Thanks for the instructions. I have turned the hooks DEBUG on, but there is no bug anymore after I restart my machine. 

This PR passed the snap testing CI, and I have also tested it manually.

File: snap/hooks/configure [hunk 6]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash -e
+
+$SNAP/bin/helper-go configure
```
Comment (farshidtz) [inline snap/hooks/configure]:
```suggestion
exec $SNAP/bin/helper-go configure
```
So that helper-go is executed as the same process rather than a child process of configure bash script. Snap wrapper scripts usually do that.

Same change for the install script.

**Comment by MonicaisHer** [comment]
@farshidtz This is ready for review. Should the further review be here, or should I create an upstream PR now?

**Comment by farshidtz** [comment]
> @farshidtz Thanks for the instructions. I have turned the hooks DEBUG on, but there is no bug anymore after I restart my machine. 
> 
> This PR passed the snap testing CI, and I have also tested it manually. 

I think you had an installation before without an install / configure hook. Then you were upgrading it with a working install / configure hook. During the upgrade, install hook isn't called, so configure must have been failing due to missing directories.

**Comment by MonicaisHer** [comment]
Upstream PR's commit: https://github.com/edgexfoundry/device-virtual-go/pull/290/commits/d1421926db38bb431c7d6cdc347c63bca0439166
